src/ast/SemanticChecksVisitor.java

```diff
@@ -35,32 +42,294 @@ public void visit(Program program) {
 
     @Override
     public void visit(ClassDecl classDecl) {
+        if(classDecl.superName() != null){
+            String superName = classDecl.superName();
+            //semantic check - no variable with name already defined in ancestor (req 4)
+            Set<String> parentFieldNames = new HashSet<>();
+            for(STSymbol symbol : instanceTemplates.get(superName)) parentFieldNames.add(symbol.name());
+            for(VarDecl varDecl : classDecl.fields()){
+                if(parentFieldNames.contains(varDecl.name())){
+                    visitResult = false;
+                    return;
+                }
+            }
+
+            //semantic check - overriding methods match signature (req 6)
+            Map<String, MethodDecl> parentMethodDecls = new HashMap<>();
+            for(STSymbol symbol : vtables.get(superName)){
+                parentMethodDecls.put(symbol.name(),(MethodDecl) symbol.declaration() );
+            }
```

Comment:
This can be simplified by calling keySet:
https://docs.oracle.com/javase/8/docs/api/java/util/HashMap.html#keySet--